### PR TITLE
Issue #13038: handled inner class method definition in VariableDeclarationUsageDistanceCheck

### PR DESCRIPTION
**Pull request by jxr98**
Issue: #13038 
Diff Regression config: https://gist.githubusercontent.com/jxr98/8f42477fea778a2b62873c9af6ad8493/raw/b6b8b98a74810181a66fc2cdb04dfc7f45109f85/VariableDeclarationUsageDistance.xml

### CODE REVIEW

**Comment by jxr98** [comment]
GitHub, generate report

**Comment by jxr98** [comment]
I've read the pitest wiki and the pitest mutation overview. I understand `removed conditional - replaced equality check with true` mutation, but don't see how it applies to the [code](https://github.com/jxr98/checkstyle/blob/49133d8cc3e38f3d843cf0c3dd918bd9a2f5314e/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java#L784-L794). I have tried to replace the first `==` with equal and remove it. It doesn't seem right. Can anyone help me? Thanks.

**Comment by nrmancuso** [comment]
> I've read the pitest wiki and the pitest mutation overview. I understand `removed conditional - replaced equality check with true` mutation, but don't see how it applies to the [code](https://github.com/jxr98/checkstyle/blob/49133d8cc3e38f3d843cf0c3dd918bd9a2f5314e/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java#L784-L794). I have tried to replace the first `==` with equal and remove it. It doesn't seem right. Can anyone help me? Thanks.
> 
> image

I am fairly certain this means that now the ONLY tokens that are passed to this method are in these conditions, so we might as well return true from the isZeroDistanceToken method, which means that where ever this method is used could probably be mutated to just `true` as well. Try to remove this method completely and see what happens.

**Comment by jxr98** [comment]
> > I've read the pitest wiki and the pitest mutation overview. I understand `removed conditional - replaced equality check with true` mutation, but don't see how it applies to the [code](https://github.com/jxr98/checkstyle/blob/49133d8cc3e38f3d843cf0c3dd918bd9a2f5314e/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java#L784-L794). I have tried to replace the first `==` with equal and remove it. It doesn't seem right. Can anyone help me? Thanks.
> > image
> 
> I am fairly certain this means that now the ONLY tokens that are passed to this method are in these conditions, so we might as well return true from the isZeroDistanceToken method, which means that where ever this method is used could probably be mutated to just `true` as well. Try to remove this method completely and see what happens.

I've tried to return true or false in the isZeroDistanceToken method and remove it completely, but some of the tests we have failed after that.

**Comment by romani** [comment]
as a hack, please make method as public and write junit test for this method to clearly show that survival can gone if we do have explicit test.

**Comment by jxr98** [comment]
> as a hack, please make method as public and write junit test for this method to clearly show that survival can gone if we do have explicit test.

Thanks. The method is very straightforward containing several or operator as follows 
image, and the code is [here](https://github.com/jxr98/checkstyle/blob/49133d8cc3e38f3d843cf0c3dd918bd9a2f5314e/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java#L784-L794). From debugging and experience, the survivals are `Expr`, `if`, `for`, `return`, `{`, etc. The conditions in the method can't be complete I guess. 🤔

**Comment by nrmancuso** [comment]
> but some of the tests we have failed after that. Still can't figure out why

If this is true, then we are missing tests from the pitest profile in pom.xml for this check. Please push up your changes and show this failure so that we can help debug.

**Comment by jxr98** [comment]
> > but some of the tests we have failed after that. Still can't figure out why
> 
> If this is true, then we are missing tests from the pitest profile in pom.xml for this check. Please push up your changes and show this failure so that we can help debug.

I've pushed the changes I made. If the method just return true directly, the failed tests are shown as in the file below. 
[com.puppycrawl.tools.checkstyle.checks.coding.VariableDeclarationUsageDistanceCheckTest.txt](https://github.com/checkstyle/checkstyle/files/14843742/com.puppycrawl.tools.checkstyle.checks.coding.VariableDeclarationUsageDistanceCheckTest.txt)

It looks like the pom file includes the target class and target test of this check correctly. I'd really appreciate it if you could help me debug. I guess I might make some mistakes where I'm not aware of right now.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceGeneral4.java [hunk 1]
```diff
@@ -22,8 +22,7 @@ void mm() {
         <T> void xx(List<T> m){}
     }
     public void method9() {
-        // until https://github.com/checkstyle/checkstyle/issues/13011
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceGeneral4.java:25]: First I would confirm that these 2 issues are exactly the same and explain it so we understand. I find it odd that a method issue would fix a class issue, but maybe no one noticed the cases were extremely similar.

Also, apparently you reported both issues. :)

**Comment by nrmancuso** [comment]
> It looks like the pom file includes the target class and target test of this check correctly. I'd really appreciate it if you could help me debug. I guess I might make some mistakes where I'm not aware of right now.

There is definitely something weird going on here. I've checked this code out locally, and tried to hardcode the mutation in a number of ways, but the tests are always failing. The only thing I can think of is that the bytecode inlines this method and the actual mutation isn't what is shown in the suppression. 

I suggest that you check the bytecode for this file, you can see how to do this at https://github.com/checkstyle/checkstyle/wiki/How-to-Generate-and-Analyze-Pitest-Reports#decompiler-usage-to-see-bytecode.

@Vyom-Yadav have you seen anything like this before?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java [hunk 2]
```diff
@@ -784,7 +787,8 @@ private static boolean isZeroDistanceToken(int type) {
                 || type == TokenTypes.MODIFIERS
                 || type == TokenTypes.RESOURCE
                 || type == TokenTypes.EXTENDS_CLAUSE
-                || type == TokenTypes.IMPLEMENTS_CLAUSE;
+                || type == TokenTypes.IMPLEMENTS_CLAUSE
+                || type == TokenTypes.CLASS_DEF;
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:791]: > removed conditional - replaced equality check with true

This is my local when I run the profile.
Image

I removed one of these lines completely, or something similar to changing `|| type == TokenTypes.EXTENDS_CLAUSE` to `|| false`, and all the tests passed. This shows the mutation. The mutation in this case is reversed from the text that says true.

So now we must identify how to kill these mutations, or understand why they can't be killed. I would look over https://github.com/checkstyle/checkstyle/wiki/How-to-Generate-and-Analyze-Pitest-Reports#how-to-find-java-code-that-can-kill-mutation and https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#checkstyle-pitest-regression for the next steps. You can also look over other pitest issues that describe the more complex process. We do not normally accept new code with new surviving mutations.

Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:791]: >At a glance this sort of makes sense. extends and implements both have class as a parent.

Yes, this sounds like what is happening here. We should probably add some test cases for local enums, records, and interfaces as well.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceGeneral4.java (hunk 1 above)
Comment (jxr98) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/variabledeclarationusagedistance/InputVariableDeclarationUsageDistanceGeneral4.java:25]: Oo right they r similar.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java (hunk 2 above)
Comment (jxr98) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/VariableDeclarationUsageDistanceCheck.java:791]: Thank u so much!!! I see. Trying to solve it.

**Comment by romani** [comment]
@jxr98, please find time to finish PR

**Comment by jxr98** [comment]
Okay sry a bit busy recently. Will finish soon.

**Comment by romani** [comment]
@MohanadKh03 , please help us here .
We appreciate your help.

**Comment by MohanadKh03** [comment]
@romani Sorry for the late reply , Sure 
It seems the issue itself is fixed and the only thing remaining is resolving the mutation right ?

**Comment by romani** [comment]
We don't know :)

**Comment by romani** [comment]
@MohanadKh03, I hope you are doing well. 
Please let me know I can answer any your questions.
We appreciate your help on this PR.

**Comment by MohanadKh03** [comment]
@romani Hey thanks for the ping I am doing fine just got a bit busy
ill be able to go through this PR next week or sth but Ill see if there is another issue/PR hanging that wouldnt need much pitest resolution such as this one and go through it

**Comment by romani** [comment]
Thanks a lot for response, if restore PR, I can help with pitest coverage ideas.
Together we will be able to finish it.

**Comment by MohanadKh03** [comment]
So after rebasing locally and doing some investigations it seems that the issue is fixed but the mutations are caused due to these conditions 
```java
|| type == TokenTypes.IMPLEMENTS
|| type == TokenTypes.EXTENDS
|| type == TokenTypes.CLASS_DEF
```
so as far as I can see what is causing the problem is that if the token type is a class then it 100% is a class therefore there will always be a mutation in these conditions
as @nrmancuso mentioned in an earlier comment we probably need to add tests for some inner classes , records and hopefully the mutations will be killed but I do wonder do we need to add another condition for records ? maybe idk

Ill try all this out in a new PR as an extension of this one alright ? @romani Also if there is something I am missing about my conclusion please let me know if there is sth else that is probably causing this mutation

**Comment by nrmancuso** [comment]
@MohanadKh03 thanks for your analysis :) I find that when working on Checkstyle, test driven development is a great way to approach most issues. Take the time to craft some interesting input files and tests, and let them guide your changes.

**Comment by romani** [comment]
Please create PR, we will discuss all in it.
We might be even luck to find case by regression.

**Comment by romani** [comment]
>wonder do we need to add another condition for records ? maybe idk

Just drop it (record ) to Input file and lets see how it works.

There more tests the better for project.

**Comment by romani** [comment]
Energy is moved to https://github.com/checkstyle/checkstyle/pull/16026

**Pull request closed without merging**